### DISCUSSION TITLE
How to reimplement operators on TypeDecorator subclasses

### DISCUSSION THREAD

**Discussion by R0land013** [sqlalchemy/sqlalchemy#8355] [Usage Questions] [answered]
I created a custom column to store money in a sqlite database.

``` Python

class MoneyColumn(TypeDecorator):

    impl = Text

    cache_ok = True

    def process_bind_param(self, money: Money, dialect):
        return '{} {}'.format(money.currency, money.amount)

    def process_result_value(self, money_as_str: str, dialect) -> Money:
        if money_as_str is None:
            return None
        money_as_list = money_as_str.split()
        currency = money_as_list[0]
        amount = money_as_list[1]
        return Money(amount, currency)

    def copy(self, **kw):
        return MoneyColumn()
```

The Money class I am using here is from the [Carlos Palol's package, called money](https://pypi.org/project/money/).
I store the money as a string in this way `'<currency> <amount>'`.

I have a problem when I execute a query with an ORDER BY clause on this money column. The ordering is not considering the values in this column as Money type, furthermore the ordering is executing string comparisons to sort the values instead of Money comparisons. In this way the sort returns wrong results.

An example of wrong behaviour is when the next values are stored:

- 'USD 40.00'
- 'USD 5.00'

If I try to sort the values in descending way I will get the next result:

- 'USD 5.00'
- 'USD 40.00'

How could I achieve the correct behaviour?

**Comment by zzzeek** [comment, marked as answer]
well ideally you'd store currency and amount in two separate columns.     It's not clear how you want to approach ordering when two different rows are in different currencies also?  how did you want to do that part?

Otherwise, you need to ORDER on a SQL function that can extract from the string the numeric value you want.  Most databases have a substr() function of some kind and then you would CAST that to Numeric, like:

    stmt.order_by(cast(func.substr(table.c.col, 3, 10), Numeric))

you could add an operator to your custom type called "as_numeric" or something like that which produces this SQL expression.

**Reply by zzzeek** [reply]
but anyway, that such a function is needed in the first place indicates you are having to do work to extract data from your columns, which indicates that these should really be two separate columns with the most atomic level of data each.

**Reply by R0land013** [reply]
Hello. Thanks for your answer. You are right. Using a column for the amount is a better design. I have changed to a column with the amount.
I resolved my issue, thank you very much.
